### PR TITLE
Add support for watch providers

### LINKED ISSUE #240: Missing Watch Providers Support
There is currently no support for the TMDb API's "Watch Providers" data.

- Movie > Watch Providers
- TV Show > Watch Providers
- Movie Discover > Filter by Watch Providers
- TV Show Discover > Filter by Watch Providers
- Watch Providers > Available Regions
- Watch Providers > Movie Providers
- Watch Providers > TV Show Providers

### PR DESCRIPTION
**Pull request by neildaniels**
This is an attempt to resolve #240 by adding support for Watch Providers.

Feedback, patches, or alternative PRs more than welcome.

- [x] Movie > Watch Providers
- [x] TV Show > Watch Providers
- [x] Movie Discover > Filter by Watch Providers
- [x] TV Show Discover > Filter by Watch Providers
- [ ] Watch Providers > Available Regions
- [ ] Watch Providers > Movie Providers
- [ ] Watch Providers > TV Show Providers

### CODE REVIEW

File: lib/Tmdb/Factory/MovieFactory.php [hunk 1]
```diff
@@ -206,6 +207,33 @@ public function create(array $data = []): ?AbstractModel
             $movie->setReleaseDates($release_dates);
         }
 
+        if (array_key_exists('watch/providers', $data) && array_key_exists('results', $data['watch/providers'])) {
+            $watchProviders = new GenericCollection();
+            foreach ($data['watch/providers']['results'] as $iso_31661 => $country_watch_data) {
```
Comment (wtfzdotnet) [inline lib/Tmdb/Factory/MovieFactory.php]: I'd prefer seeing camel-cased variables :-)

File: lib/Tmdb/Model/Watch/Provider.php [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+/**
+ * This file is part of the Tmdb PHP API created by Michael Roterman.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ *
+ * @package Tmdb
+ * @author Michael Roterman <michael@wtfz.net>
+ * @copyright (c) 2013, Michael Roterman
```
Comment (wtfzdotnet) [inline lib/Tmdb/Model/Watch/Provider.php]: Credits go to you here @neildaniels :-)

File: lib/Tmdb/Model/Watch/Provider.php [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    {
+        $this->iso31661 = $iso31661;
+
+        return $this;
+    }
+
+    /**
+     * @return int|null
+     */
+    public function getID()
```
Comment (wtfzdotnet) [inline lib/Tmdb/Model/Watch/Provider.php]: is the capitalization consistent with the library?

File: lib/Tmdb/Model/Watch/Provider.php [hunk 4]
```diff
@@ -0,0 +1,154 @@
+<?php
+
```
Comment (wtfzdotnet) [inline lib/Tmdb/Model/Watch/Provider.php:2]: Not sure if there's anything in the way to use strict type hinting within the library by now? Altough not everything is aligned like that, I'd sure like to see that fixed in time.

File: lib/Tmdb/Model/Watch/Providers.php [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+/**
+ * This file is part of the Tmdb PHP API created by Michael Roterman.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ *
+ * @package Tmdb
+ * @author Michael Roterman <michael@wtfz.net>
+ * @copyright (c) 2013, Michael Roterman
```
Comment (wtfzdotnet) [inline lib/Tmdb/Model/Watch/Providers.php]: Same here, your credits @neildaniels

File: lib/Tmdb/Model/Watch/Providers.php [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    public function getFlatrate()
+    {
+        return $this->flatrate;
+    }
+
+    /**
+     * @param GenericCollection $flatrate
+     * @return $this
+     */
+    public function setFlatrate($flatrate)
```
Comment (wtfzdotnet) [inline lib/Tmdb/Model/Watch/Providers.php]: Could use the typehint for `GenericCollection`

File: test/Tmdb/Tests/Repository/MovieRepositoryTest.php [hunk 7]
```diff
@@ -30,7 +30,7 @@ public function shouldLoadMovie()
         $repository->load(self::MOVIE_ID);
         $this->assertLastRequestIsWithPathAndMethod('/3/movie/' . self::MOVIE_ID);
         $this->assertRequestHasQueryParameters([
-            'append_to_response' => 'alternative_titles,external_ids,changes,credits,images,keywords,lists,release_dates,reviews,similar,recommendations,translations,videos'
+            'append_to_response' => 'alternative_titles,external_ids,changes,credits,images,keywords,lists,release_dates,reviews,similar,recommendations,translations,videos,watch/providers'
```
Comment (wtfzdotnet) [inline test/Tmdb/Tests/Repository/MovieRepositoryTest.php:33]: Any chance you could update the stored data ( flat json files ) that include this data and add some specific tests? :-)

**Comment by neildaniels** [comment]
This PR still does not include support for fetching the list of available regions, list of movie providers, and list of tv providers.

However, I have added tests for the behavior it does support. I think this PR is still worth merging in, as I'm unsure when I'd have time to add support for the remaining parts.

**Comment by wtfzdotnet** [comment]
> This PR still does not include support for fetching the list of available regions, list of movie providers, and list of tv providers.
> 
> However, I have added tests for the behavior it does support. I think this PR is still worth merging in, as I'm unsure when I'd have time to add support for the remaining parts.

Checked it a few times without really getting the heads-up you wanted to merge it without all tasks completed; going to check it now :-).

**Review by wtfzdotnet — APPROVED** [review]

**Comment by wtfzdotnet** [comment]
Thanks for the addition @neildaniels  :-), cheers!

**Pull request merged**